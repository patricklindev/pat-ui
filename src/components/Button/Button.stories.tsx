import React from 'react';
import { action } from '@storybook/addon-actions';
import Button from './Button';

export default {
  title: 'Button',
  component: Button,
};

const buttonStyle: React.CSSProperties = {
  marginRight: '5px',
  marginTop: '5px',
};

export const DefaultButton = () => (
  <Button onClick={action('Default Button clicked')}>Default Button</Button>
);

export const DiffSizeButton = () => (
  <div>
    <Button
      style={buttonStyle}
      btnSize="sm"
      onClick={action('Small Button clicked')}
    >
      Small Button
    </Button>
    <Button style={buttonStyle} onClick={action('Default Button clicked')}>
      Default Button
    </Button>

    <Button btnSize="lg" onClick={action('Large Button clicked')}>
      Large Button
    </Button>
  </div>
);

export const DiffTypeButton = () => (
  <div>
    <Button
      style={buttonStyle}
      btnType="primary"
      onClick={action('Primary Button clicked')}
    >
      Primary Button
    </Button>
    <Button
      style={buttonStyle}
      btnType="secondary"
      onClick={action('Secondary Button clicked')}
    >
      Secondary Button
    </Button>
    <Button
      style={buttonStyle}
      btnType="danger"
      onClick={action('Danger Button clicked')}
    >
      Danger Button
    </Button>
    <Button
      style={buttonStyle}
      btnType="default"
      onClick={action('Default Button clicked')}
    >
      Default Button
    </Button>
    <Button
      style={buttonStyle}
      disabled
      btnType="default"
      onClick={action('DisabledDefault Button clicked should not work')}
    >
      Disabled Default Button
    </Button>
    <Button
      style={buttonStyle}
      btnType="link"
      onClick={action('Link Button clicked')}
    >
      Link Button
    </Button>
    <Button
      disabled
      btnType="link"
      onClick={action('Disabled Link Button clicked should not work')}
    >
      Disabled Link Button
    </Button>
  </div>
);
