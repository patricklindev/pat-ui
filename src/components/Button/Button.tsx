import React, {
  ButtonHTMLAttributes,
  AnchorHTMLAttributes,
  FC,
  MouseEvent,
} from 'react';
import { classNames } from '../../utils/classNames';

// export enum ButtonSize {
//   Large = 'lg',
//   Small = 'sm',
// }
export type ButtonSize = 'lg' | 'sm';
export type ButtonType =
  | 'primary'
  | 'secondary'
  | 'danger'
  | 'default'
  | 'link';
// export enum ButtonType {
//   Primary = 'primary',
//   Secondary = 'secondary',
//   Danger = 'danger',
//   Default = 'default',
//   Link = 'link',
// }
export interface IButtonProps {
  /** set customized style */
  className?: string;
  /** set button size */
  btnSize?: ButtonSize;
  /** set button type */
  btnType?: ButtonType;
  /** set disabled button */
  disabled?: boolean;
}

type NativeButtonProps = IButtonProps & ButtonHTMLAttributes<HTMLButtonElement>;
type NativeAchorButtonProps = IButtonProps &
  AnchorHTMLAttributes<HTMLAnchorElement>;
export type PatButtonProps = NativeButtonProps | NativeAchorButtonProps;

/**
 * A Button indicates a possible user action.
 *
 * ```js
 * import {Button} from 'pat-ui'
 * ```
 */
export const Button: FC<PatButtonProps> = (props) => {
  const { btnSize, btnType, children, disabled, className, ...rest } = props;
  let styleClasses = classNames('btn', {
    [`btn-${btnType}`]: true,
    [`btn-${btnSize}`]: !!btnSize,
    disabled: !!(disabled && btnType === 'link'),
  });
  if (className) {
    styleClasses += ' ' + className;
  }

  let btn;
  if (btnType !== 'link') {
    btn = (
      <button
        className={styleClasses}
        disabled={disabled}
        {...(rest as NativeButtonProps)}
      >
        {props.children + 'test'}
      </button>
    );
  } else {
    if (disabled) {
      rest.onClick = (e: MouseEvent) => {
        e.preventDefault();
      };
    }
    btn = (
      <a className={styleClasses} {...(rest as NativeAchorButtonProps)}>
        {props.children + 'test'}
      </a>
    );
  }
  return btn;
};

Button.defaultProps = {
  btnType: 'default',
  disabled: false,
};

export default Button;
