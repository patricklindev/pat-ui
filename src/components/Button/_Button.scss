.btn {
  position: relative;
  display: inline-block;
  font-weight: $btn-font-weight;
  line-height: $btn-line-height;
  color: $body-color;
  white-space: nowrap;
  text-align: center;
  vertical-align: middle;
  border: $btn-border-width solid transparent;
  box-shadow: $btn-box-shadow;
  cursor: pointer;
  transition: $btn-transition;

  // btn size
  @include button-size(
    $btn-padding-y,
    $btn-padding-x,
    $btn-font-size,
    $border-radius
  );

  &.disabled,
  &[disabled] {
    cursor: not-allowed;
    opacity: $btn-disabled-opacity;
    box-shadow: none;
  }
}

.btn-sm {
  @include button-size(
    $btn-padding-y-sm,
    $btn-padding-x-sm,
    $btn-font-size-sm,
    $border-radius-sm
  );
}
.btn-lg {
  @include button-size(
    $btn-padding-y-lg,
    $btn-padding-x-lg,
    $btn-font-size-lg,
    $border-radius-lg
  );
}

.btn-default {
  @include button-style(
    $body-color,
    $white,
    $gray-400,
    $primary,
    $white,
    $primary
  );
}

.btn-primary {
  @include button-style($white, $primary, $primary);
}
.btn-secondary {
  @include button-style($white, $secondary, $secondary);
}
.btn-danger {
  @include button-style($white, $danger, $danger);
}

.btn-link {
  font-weight: $font-weight-normal;
  color: $btn-link-color !important;
  text-decoration: $link-decoration;
  box-shadow: none;
  &:hover {
    color: $btn-link-hover-color;
    text-decoration: $link-hover-decoration;
  }
  &:focus,
  &.focus {
    box-shadow: none;
  }
}
