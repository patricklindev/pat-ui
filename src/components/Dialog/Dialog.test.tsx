import React from 'react';

import { render, fireEvent } from '@testing-library/react';
import Button, { PatButtonProps } from './Dialog';

describe('Buttton', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<Button> Snapshot Button </Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render default button', () => {
    const btnProps = {
      onClick: jest.fn(),
    };
    const wrapper = render(<Button {...btnProps}>Default Button</Button>);
    const element = wrapper.queryByText('Default Button') as HTMLElement;
    expect(element).toBeInTheDocument();
    expect(element.tagName).toBe('BUTTON');
    expect(element).toHaveClass('btn btn-default');
    expect(btnProps.onClick).toHaveBeenCalledTimes(0);
    fireEvent.click(element);
    expect(btnProps.onClick).toHaveBeenCalledTimes(1);
  });

  it('should render correct button based on different props', () => {
    const btnPrimarySmallProps: PatButtonProps = {
      btnType: 'primary',
      btnSize: 'sm',
      onClick: jest.fn(),
      className: 'test',
    };
    const btnPrimarySmallWrapper = render(
      <Button {...btnPrimarySmallProps}>Primary Small Button</Button>
    );
    const btnPrimarySmallElement = btnPrimarySmallWrapper.queryByText(
      'Primary Small Button'
    ) as HTMLElement;
    expect(btnPrimarySmallElement).toBeInTheDocument();
    expect(btnPrimarySmallElement.tagName).toBe('BUTTON');
    expect(btnPrimarySmallElement).toHaveClass('btn btn-primary btn-sm test');
    expect(btnPrimarySmallProps.onClick).toHaveBeenCalledTimes(0);
    fireEvent.click(btnPrimarySmallElement);
    expect(btnPrimarySmallProps.onClick).toHaveBeenCalledTimes(1);

    /// Link Button
    const btnLinkProps: PatButtonProps = {
      btnType: 'link',
      onClick: jest.fn(),
    };
    const btnLinkWrapper = render(
      <Button {...btnLinkProps}>Link Button</Button>
    );
    const btnLinkElement = btnLinkWrapper.queryByText(
      'Link Button'
    ) as HTMLElement;
    expect(btnLinkElement).toBeInTheDocument();
    expect(btnLinkElement.tagName).toBe('A');
    expect(btnLinkElement).toHaveClass('btn btn-link');
    expect(btnLinkProps.onClick).toHaveBeenCalledTimes(0);
    fireEvent.click(btnLinkElement);
    expect(btnLinkProps.onClick).toHaveBeenCalledTimes(1);
  });

  it('should render disabled button', () => {
    const btnDisabledLinkProps: PatButtonProps = {
      btnType: 'link',
      onClick: jest.fn(),
      disabled: true,
    };
    const btnDisabledLinkWrapper = render(
      <Button {...btnDisabledLinkProps}>Disabled Link Button</Button>
    );
    const btnDisabledLinkElement = btnDisabledLinkWrapper.queryByText(
      'Disabled Link Button'
    ) as HTMLElement;
    expect(btnDisabledLinkElement).toBeInTheDocument();
    expect(btnDisabledLinkElement.tagName).toBe('A');
    expect(btnDisabledLinkElement).toHaveClass('btn btn-link disabled');
    expect(btnDisabledLinkProps.onClick).toHaveBeenCalledTimes(0);
    fireEvent.click(btnDisabledLinkElement);
    expect(btnDisabledLinkProps.onClick).toHaveBeenCalledTimes(0);

    const btnDisabledProps = {
      onClick: jest.fn(),
      disabled: true,
    };
    const btnDisabledWrapper = render(
      <Button {...btnDisabledProps}>Disabled Button</Button>
    );
    const btnDisabledElement = btnDisabledWrapper.queryByText(
      'Disabled Button'
    ) as HTMLElement;
    expect(btnDisabledElement).toBeInTheDocument();
    expect(btnDisabledElement.tagName).toBe('BUTTON');
    expect(btnDisabledElement).toHaveClass('btn btn-default');
    expect(btnDisabledElement).not.toHaveClass('disalbed');
    expect(btnDisabledProps.onClick).toHaveBeenCalledTimes(0);
    fireEvent.click(btnDisabledElement);
    expect(btnDisabledProps.onClick).toHaveBeenCalledTimes(0);
  });
});