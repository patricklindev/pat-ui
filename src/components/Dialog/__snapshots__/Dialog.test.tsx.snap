// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttton should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-default"
  >
     Snapshot Button 
  </button>
</DocumentFragment>
`;
